### PR TITLE
KAFKA-15237: Implement write operation timeout

### PR DESCRIPTION
**Pull request by vamossagar12**
Similar to the old Group Coordinator, this PR tries to bound all writes on the `offset.commit.timeout.ms` config to ensure that the number of messages in the purgatory remains bounded.

### CODE REVIEW

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 1]
```diff
@@ -595,22 +609,25 @@ class CoordinatorWriteEvent<T> implements CoordinatorEvent, DeferredEvent {
...
+         * @param name      The operation name.
+         * @param tp        The topic partition that the operation is applied to.
+         * @param op        The write operation.
+         * @param timeout   The write operation timeout
          */
         CoordinatorWriteEvent(
             String name,
             TopicPartition tp,
-            CoordinatorWriteOperation<S, T, U> op
+            CoordinatorWriteOperation<S, T, U> op,
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: It would be better to keep the `op` as the last argument.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 2]
```diff
@@ -623,14 +640,16 @@ class CoordinatorWriteEvent<T> implements CoordinatorEvent, DeferredEvent {
...
+         * @param timeout           The write operation timeout
          */
         CoordinatorWriteEvent(
             String name,
             TopicPartition tp,
             String transactionalId,
             long producerId,
             short producerEpoch,
-            CoordinatorWriteOperation<S, T, U> op
+            CoordinatorWriteOperation<S, T, U> op,
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: ditto.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 3]
```diff
@@ -581,6 +589,12 @@ class CoordinatorWriteEvent<T> implements CoordinatorEvent, DeferredEvent {
          */
         final CompletableFuture<T> future;
 
+        /**
+         * Timeout value for the write operation
+         */
+
+        final int timeout;
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: I wonder if we should rather use a Duration. What do you think?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 4]
```diff
@@ -698,6 +718,14 @@ public void run() {
                                 producerEpoch,
                                 result.records()
                             );
+
+                            timer.add(new TimerTask(timeout) {
+                                @Override
+                                public void run() {
+                                    scheduleInternalOperation(name, tp, () -> complete(new TimeoutException("Writing records to the log timed out")));
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: I wonder if we could more details in the exception (e.g. event name, etc.).

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 5]
```diff
@@ -1229,11 +1266,12 @@ private void withActiveContextOrThrow(
     public <T> CompletableFuture<T> scheduleWriteOperation(
         String name,
         TopicPartition tp,
-        CoordinatorWriteOperation<S, T, U> op
+        CoordinatorWriteOperation<S, T, U> op,
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: nit: Let's keep `op` as the last param.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 6]
```diff
@@ -1259,7 +1298,8 @@ public <T> CompletableFuture<T> scheduleTransactionalWriteOperation(
         String transactionalId,
         long producerId,
         short producerEpoch,
-        CoordinatorWriteOperation<S, T, U> op
+        CoordinatorWriteOperation<S, T, U> op,
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: ditto.

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java [hunk 7]
```diff
@@ -950,6 +948,55 @@ public void testScheduleWriteOpWhenWriteFails() {
...
+
+        // Verify the initial state.
+        CoordinatorRuntime<MockCoordinatorShard, String>.CoordinatorContext ctx = runtime.contextOrThrow(TP);
+        assertEquals(0, ctx.lastWrittenOffset);
+        assertEquals(0, ctx.lastCommittedOffset);
+        assertEquals(Collections.singletonList(0L), ctx.snapshotRegistry.epochsList());
+
+        // The partition writer times out when trying to append records.
+        doAnswer(invocation -> {
+            time.sleep(DEFAULT_WRITE_TIMEOUT * 3);
```
Comment (dajac) [inline group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java]: You actually don't need this because the timeout will be reached if the HWM is not advanced.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 8]
```diff
@@ -92,6 +93,7 @@ public static class Builder<S extends CoordinatorShard<U>, U> {
         private CoordinatorShardBuilderSupplier<S, U> coordinatorShardBuilderSupplier;
         private Time time = Time.SYSTEM;
         private Timer timer;
+        private int timeout;
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: nit: should we call it `defaultWriteTimeout`?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java (hunk 4 above)
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: I also wonder whether we could cancel the timer or skip scheduling the internal op if when the event is completed. I can think of two ways: 1) We keep a reference to TimerTask and call `cancel` on it; 2) We check the `future` before calling `scheduleInternalOperation`.

**Review by dajac — COMMENTED** [review]
@vamossagar12 Thanks for the PR. I left some comments.

**Comment by dajac** [comment]
@vamossagar12 I updated the title to match the one of the Jira.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 9]
```diff
@@ -719,20 +721,26 @@ public void run() {
...
+                                    }
                                 }
                             });
 
-                            context.coordinator.updateLastWrittenOffset(offset);
-
                             // Add the response to the deferred queue.
                             if (!future.isDone()) {
                                 context.deferredEventQueue.add(offset, this);
+                                context.coordinator.updateLastWrittenOffset(offset);
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: Why did you move this? I think that it should stay where it was and the timer must be added here, no?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 10]
```diff
@@ -135,6 +138,11 @@ public Builder<S, U> withTimer(Timer timer) {
             return this;
         }
 
+        public Builder<S, U> withWriteTimeOut(Duration timeout) {
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: nit: `withDefaultWriteTimeout`?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 11]
```diff
@@ -581,6 +591,12 @@ class CoordinatorWriteEvent<T> implements CoordinatorEvent, DeferredEvent {
          */
         final CompletableFuture<T> future;
 
+        /**
+         * Timeout value for the write operation
+         */
+
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: nit: Remove empty line.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 12]
```diff
@@ -581,6 +591,12 @@ class CoordinatorWriteEvent<T> implements CoordinatorEvent, DeferredEvent {
...
+        final Duration timeout;
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java:598]: Should it be `defaultWriteTimeout` too?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 13]
```diff
@@ -595,13 +611,15 @@ class CoordinatorWriteEvent<T> implements CoordinatorEvent, DeferredEvent {
...
-         * @param op    The write operation.
+         * @param name      The operation name.
+         * @param tp        The topic partition that the operation is applied to.
+         * @param timeout   The write operation timeout
+         * @param op        The write operation.
          */
         CoordinatorWriteEvent(
             String name,
             TopicPartition tp,
+            Duration timeout,
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java:622]: Same question about the name.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 14]
```diff
@@ -698,13 +720,27 @@ public void run() {
...
+                                }
+                            });
 
                             // Add the response to the deferred queue.
                             if (!future.isDone()) {
                                 context.deferredEventQueue.add(offset, this);
+                                context.coordinator.updateLastWrittenOffset(offset);
                             } else {
                                 complete(null);
+
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: nit: We could remove this empty line.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 15]
```diff
@@ -1019,6 +1055,11 @@ public void onHighWatermarkUpdated(
      */
     private final Timer timer;
 
+    /**
+     * The write operation timeout
+     */
+    private final Duration timeout;
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: nit: `defaultWriteTimeout`?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 16]
```diff
@@ -1092,6 +1134,7 @@ private CoordinatorRuntime(
         CoordinatorShardBuilderSupplier<S, U> coordinatorShardBuilderSupplier,
         Time time,
         Timer timer,
+        Duration timeout,
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: nit `defaultWriteTimeout`?

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupCoordinatorServiceTest.java [hunk 17]
```diff
@@ -186,6 +187,7 @@ public void testConsumerGroupHeartbeat() throws ExecutionException, InterruptedE
         when(runtime.scheduleWriteOperation(
             ArgumentMatchers.eq("consumer-group-heartbeat"),
             ArgumentMatchers.eq(new TopicPartition("__consumer_offsets", 0)),
+            ArgumentMatchers.any(),
```
Comment (dajac) [inline group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupCoordinatorServiceTest.java]: Could we actually put the expected value for all of those?

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java [hunk 18]
```diff
@@ -950,6 +965,43 @@ public void testScheduleWriteOpWhenWriteFails() {
         assertEquals(mkSet("record1", "record2"), ctx.coordinator.coordinator().records());
     }
 
+    @Test
+    public void testScheduleWriteOpWhenWriteTimesout() throws InterruptedException {
```
Comment (dajac) [inline group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java]: nit: `TimesOut`?

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java [hunk 19]
```diff
@@ -950,6 +965,43 @@ public void testScheduleWriteOpWhenWriteFails() {
...
+        // Loads the coordinator.
+        runtime.scheduleLoadOperation(TP, 10);
+
+        // Verify the initial state.
+        CoordinatorRuntime<MockCoordinatorShard, String>.CoordinatorContext ctx = runtime.contextOrThrow(TP);
+        assertEquals(0, ctx.coordinator.lastWrittenOffset());
+        assertEquals(0, ctx.coordinator.lastCommittedOffset());
+        assertEquals(Collections.singletonList(0L), ctx.coordinator.snapshotRegistry().epochsList());
+
+        // Write #1. We should get a TimeoutException because the HWM will not advance
```
Comment (dajac) [inline group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java]: nit: `.` at the end.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java (hunk 9 above)
Comment (vamossagar12) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: Ok. yes, that was a mistake.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 20]
```diff
@@ -135,6 +138,11 @@ public Builder<S, U> withTimer(Timer timer) {
             return this;
         }
 
+        public Builder<S, U> withDefaultWriteTimeOut(Duration timeout) {
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: nit: `timeout` -> `defaultWriteTimeout`.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 21]
```diff
@@ -581,6 +590,11 @@ class CoordinatorWriteEvent<T> implements CoordinatorEvent, DeferredEvent {
          */
         final CompletableFuture<T> future;
 
+        /**
+         * Timeout value for the write operation
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: nit: `.`.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 22]
```diff
@@ -581,6 +590,11 @@ class CoordinatorWriteEvent<T> implements CoordinatorEvent, DeferredEvent {
...
+         */
+        final Duration defaultWriteTimeout;
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: nit: This one should actually be `writeTimeout`.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 23]
```diff
@@ -595,13 +609,15 @@ class CoordinatorWriteEvent<T> implements CoordinatorEvent, DeferredEvent {
...
-         * @param op    The write operation.
+         * @param name                  The operation name.
+         * @param tp                    The topic partition that the operation is applied to.
+         * @param defaultWriteTimeout   The default write operation timeout
+         * @param op                    The write operation.
          */
         CoordinatorWriteEvent(
             String name,
             TopicPartition tp,
+            Duration defaultWriteTimeout,
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: ditto.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 24]
```diff
@@ -610,26 +626,29 @@ class CoordinatorWriteEvent<T> implements CoordinatorEvent, DeferredEvent {
...
+         * @param defaultWriteTimeout       The write operation timeout
+         * @param op                        The write operation.
          */
         CoordinatorWriteEvent(
             String name,
             TopicPartition tp,
             String transactionalId,
             long producerId,
             short producerEpoch,
+            Duration defaultWriteTimeout,
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: ditto.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 25]
```diff
@@ -703,6 +723,18 @@ public void run() {
                             // Add the response to the deferred queue.
                             if (!future.isDone()) {
                                 context.deferredEventQueue.add(offset, this);
+                                timer.add(new TimerTask(defaultWriteTimeout.toMillis()) {
+                                    @Override
+                                    public void run() {
+                                        if (!future.isDone()) {
+                                            scheduleInternalOperation(
+                                                "LogAppendEvent(name=" + name + ", tp=" + tp + ")",
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: nit: `WriteTimeout`?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java [hunk 26]
```diff
@@ -703,6 +723,18 @@ public void run() {
...
+                                                tp,
+                                                () -> complete(new TimeoutException("Log append event " + name + "timed out for TopicPartition " + tp))
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: nit: `"CoordinatorWriteEvent " + name + " timed out after " + timeoutMs + "ms"`.

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java [hunk 27]
```diff
@@ -950,6 +965,43 @@ public void testScheduleWriteOpWhenWriteFails() {
...
+        runtime.scheduleLoadOperation(TP, 10);
+
+        // Verify the initial state.
+        CoordinatorRuntime<MockCoordinatorShard, String>.CoordinatorContext ctx = runtime.contextOrThrow(TP);
+        assertEquals(0, ctx.coordinator.lastWrittenOffset());
+        assertEquals(0, ctx.coordinator.lastCommittedOffset());
+        assertEquals(Collections.singletonList(0L), ctx.coordinator.snapshotRegistry().epochsList());
+
+        // Write #1. We should get a TimeoutException because the HWM will not advance.
+        CompletableFuture<String> timedOutWrite = runtime.scheduleWriteOperation("write#1", TP, DEFAULT_WRITE_TIMEOUT,
```
Comment (dajac) [inline group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java]: nit: Could we actually use a different timeout here to ensure that the default is not used?

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java [hunk 28]
```diff
@@ -950,6 +965,43 @@ public void testScheduleWriteOpWhenWriteFails() {
...
+            state -> new CoordinatorResult<>(Arrays.asList("record1", "record2"), "response1"));
+
+        timer.advanceClock(DEFAULT_WRITE_TIMEOUT.toMillis() * 2);
```
Comment (dajac) [inline group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java]: nit: I would just do `timeout + 1` if possible.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java (hunk 26 above)
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java]: Another thing, we should use `org.apache.kafka.common.errors.TimeoutException`. Could we also ensure this in the test `testScheduleWriteOpWhenWriteTimesOut`?

**Review by dajac — COMMENTED** [review]
@vamossagar12 Thanks for the update. I left a few more minor comments.

**Comment by vamossagar12** [comment]
Thanks for the review @dajac . I have addressed the comments.

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java [hunk 29]
```diff
@@ -950,6 +965,43 @@ public void testScheduleWriteOpWhenWriteFails() {
...
+        CoordinatorRuntime<MockCoordinatorShard, String>.CoordinatorContext ctx = runtime.contextOrThrow(TP);
+        assertEquals(0, ctx.coordinator.lastWrittenOffset());
+        assertEquals(0, ctx.coordinator.lastCommittedOffset());
+        assertEquals(Collections.singletonList(0L), ctx.coordinator.snapshotRegistry().epochsList());
+
+        // Write #1. We should get a TimeoutException because the HWM will not advance.
+        CompletableFuture<String> timedOutWrite = runtime.scheduleWriteOperation("write#1", TP, Duration.ofMillis(3),
+            state -> new CoordinatorResult<>(Arrays.asList("record1", "record2"), "response1"));
+
+        timer.advanceClock(DEFAULT_WRITE_TIMEOUT.toMillis() + 1);
```
Comment (dajac) [inline group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java]: We should use `3` here to be consistent with `Duration.ofMillis(3)`.

Comment (vamossagar12) [reply group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java]: Done.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java [hunk 30]
```diff
@@ -19,16 +19,7 @@
...
-import org.apache.kafka.common.errors.InvalidFetchSizeException;
-import org.apache.kafka.common.errors.KafkaStorageException;
-import org.apache.kafka.common.errors.NotCoordinatorException;
-import org.apache.kafka.common.errors.NotEnoughReplicasException;
-import org.apache.kafka.common.errors.NotLeaderOrFollowerException;
-import org.apache.kafka.common.errors.RecordBatchTooLargeException;
-import org.apache.kafka.common.errors.RecordTooLargeException;
-import org.apache.kafka.common.errors.UnknownMemberIdException;
-import org.apache.kafka.common.errors.UnknownTopicOrPartitionException;
+import org.apache.kafka.common.errors.*;
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java]: nit: Let's revert this one. We usually don't do this.

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java [hunk 31]
```diff
@@ -997,7 +997,7 @@ public void testScheduleWriteOpWhenWriteTimesOut() throws InterruptedException {
         CompletableFuture<String> timedOutWrite = runtime.scheduleWriteOperation("write#1", TP, Duration.ofMillis(3),
             state -> new CoordinatorResult<>(Arrays.asList("record1", "record2"), "response1"));
 
-        timer.advanceClock(DEFAULT_WRITE_TIMEOUT.toMillis() + 1);
+        timer.advanceClock(DEFAULT_WRITE_TIMEOUT.toMillis() + 3);
```
Comment (dajac) [inline group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java]: nit: This should be `3 + 1`, isn't it?

**Review by dajac — APPROVED** [review]
LGTM, thanks.

**Comment by dajac** [comment]
The failed tests are not related. Merging to trunk and 3.7.
